### PR TITLE
Replace clap with our own args parser

### PR DESCRIPTION
**Pull request by lauralt**
Signed-off-by: Laura Loghin <lauralg@amazon.com>

## Reason for This PR

Get rid of `clap` crate dependencies and reduce this way the binary size.

## Description of Changes

Implements command line arguments parsing and validating and replaces `clap` crate functionality.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] No docs need to be updated as part of this PR.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in this PR have no user impact.

### CODE REVIEW

File: fc_util/src/arg_parser.rs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+        &mut self,
+        args: &'a [String],
+        args_info: &[ArgumentInfo<'static>],
+    ) -> Result<()> {
+        let seccomp_values = ["0", "1", "2"];
+        let trimmed_args: Vec<&str> = args
+            .iter()
+            .filter(|&x| x.starts_with("--"))
+            .map(|x| x.trim_start_matches("--"))
+            .collect();
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I think it's better to use `split` here. It would return a vector with all the parameter - value pairs. It might simplify things.

File: fc_util/src/arg_parser.rs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+
+        if trimmed_args.contains(&"help") {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: This logic seems a bit problematic. I think we should print this only if the input is `firecracker help`. Otherwise if for example the input is `firecracker --exec-file help`, then `help` may be the name of the file.

**Comment by lauralt** [comment]
> Why not use getopts: https://crates.io/crates/getopts?

Because it would still be an external dependency and we want to reduce the number of them if the functionality it's not that hard to replace.

File: fc_util/src/arg_parser.rs (hunk 1 above)
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: `args` is a `Vec<String>` and if I'd use `split` on it by searching for the elements that start with `--`, then `split` would remove those elements.

File: fc_util/src/arg_parser.rs (hunk 2 above)
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: Here I am searching for `help` in the `trimmed_args` list, which contains the initial arguments that started with `--` (with `--` also removed), so the search is actually for `--help`. I was thinking to print the help information regardless of  `--help` position.

File: fc_util/src/arg_parser.rs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+            }
+            if arg.starts_with("--") {
+                let name = arg.trim_start_matches("--");
+                if let Some(arg_info) = args_info.iter().find(|arg| arg.name == name) {
+                    self.check_conflicts(arg_info, &trimmed_args, index, &args)?;
+                    if arg_info.takes_value {
+                        // Safe to unwrap here because we are already checking if there is an argument
+                        // provided at `index + 1` position in `check_conflicts()` call.
+                        let arg_value = args.get(index + 1).unwrap();
+                        if name == "seccomp-level" && !seccomp_values.contains(&arg_value.as_str())
```
[the remark this answers is not in the document]
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: I agree, I've added a function for this validation in `validators.rs`.

File: fc_util/src/arg_parser.rs (hunk 2 above)
Comment (serban300) [reply fc_util/src/arg_parser.rs]: I see. Ok.

File: fc_util/src/arg_parser.rs [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+            if !trimmed_args.contains(&arg_name) {
+                return Err(Error::MissingArgument(arg_name.to_string()));
+            }
+        }
+        Ok(())
+    }
+
+    // Populates `args_matches` field of ArgParser.
+    pub fn set_matches(&mut self, args: &'a [String]) -> Result<()> {
+        let trimmed_args: Vec<&str> = args
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I think we don't need trimmed_args. I think we can do all the required checks without it and removing it could simplify the validation logic.

File: src/main.rs [hunk 5]
```diff
@@ -77,113 +80,117 @@ fn main() {
...
+            .optional(true)
+            .help("Path to a file that contains the microVM configuration in JSON format."),
+    );
+    args_parser.arg(ArgumentInfo::new(
+        "no-api")
+        .takes_value(false)
+        .requires("config-file")
+        .optional(true)
+        .help("Optional parameter which allows starting and using a microVM without an active API socket.")
+    );
```
Comment (serban300) [inline src/main.rs]: can we chain this calls to `arg` ?

File: src/main.rs [hunk 6]
```diff
@@ -77,113 +80,117 @@ fn main() {
...
+
+    // Remove the first element of `args` as it should be the binary name.
+    args.remove(0);
+    args_parser.set_matches(&args).unwrap_or_else(|err| {
```
Comment (serban300) [inline src/main.rs]: I think it would be cleaner if `set_matches` would do `env::args().collect();` inside (just like clap `get_matches`), instead of receiving the arguments when called. I would also propose to rename this method to something like `parse_cmd_line()` or `parse()`, etc.

File: fc_util/src/arg_parser.rs [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+        argument_info: &ArgumentInfo<'a>,
+        trimmed_args: &[&str],
+        index: usize,
+        args: &[String],
+    ) -> Result<()> {
+        // Check if the argument at `index` position was not provided more than once.
+        if self
+            .args_matches
+            .iter()
+            .any(|x| x.name == argument_info.name && x.user_value.is_some())
```
[the remark this answers is not in the document]
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: Done.

File: fc_util/src/arg_parser.rs [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+        // If the argument at `index` position takes a value, check if there was provided
+        // one and only one value for that argument.
+        if argument_info.takes_value {
+            if let Some(arg_value) = args.get(index + 1) {
+                if arg_value.starts_with("--")
+                    && trimmed_args.contains(&arg_value.trim_start_matches("--"))
+                {
+                    return Err(Error::MissingValue(argument_info.name.to_string()));
+                }
+                if let Some(arg) = args.get(index + 2) {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: I agree, I refactored this function to be more concise.

File: fc_util/src/arg_parser.rs (hunk 4 above)
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: Removed it.

File: src/main.rs (hunk 5 above)
Comment (lauralt) [reply src/main.rs]: Done.

File: src/main.rs (hunk 6 above)
Comment (lauralt) [reply src/main.rs]: I moved `env::args().collect()` in ArgParser implementation, but in a separate function in order to be able to test `set_maches` (which I've also renamed to `get_matches`).

File: fc_util/src/arg_parser.rs [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+            .args_matches
+            .iter()
+            .filter(|arg| arg.user_value.is_some())
+        {
+            if let Some(arg_name) = arg_info.requires {
+                if !args.contains(&(format!("--{}", arg_name))) {
+                    return Err(Error::MissingArgument(arg_name.to_string()));
+                }
+            }
+        }
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I would do both checks in a single loop.

File: fc_util/src/arg_parser.rs [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+                break;
+            }
+            if arg.starts_with("--") {
+                let name = &arg[2..];
+                if self.args_matches.get(name).is_none() {
+                    return Err(Error::UnexpectedArgument(name.to_string()));
+                }
+                // Safe to unwrap here because we've checked before if that argument exists in
+                // the list of required arguments.
+                let arg_info = self.args_matches.get(name).unwrap();
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: This can be done in just one line:

```
let arg_info = self.args_matches.get(name).ok_or(Error::UnexpectedArgument(name.to_string()))?;
```

File: fc_util/src/arg_parser.rs [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+        } else {
+            Vec::new()
+        };
+
+        // Check if the first argument provided by user doesn't start with `--`.
+        if let Some(arg) = args.get(0) {
+            if !arg.starts_with("--") {
+                return Err(Error::UnexpectedArgument(arg.clone()));
+            }
+        }
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: Is this check really necessary ? Seems covered in the following `for`.

File: fc_util/src/arg_parser.rs [hunk 12]
```diff
@@ -0,0 +1,581 @@
...
+        let mut args: Vec<String> = env::args().collect();
+        // Remove the first element of `args` as it should be the binary name.
+        args.remove(0);
+
+        self.get_matches(args)
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, args: Vec<String>) -> Result<Self> {
+        if args.contains(&"--help".to_string()) {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I would define "--" as a constant. Also "help" and all the other hardcoded strings used in the code.

File: fc_util/src/arg_parser.rs [hunk 13]
```diff
@@ -0,0 +1,581 @@
...
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Used for a general validation of the arguments provided by user.
+    fn is_valid(&self, argument_info: &ArgInfo<'a>, index: usize, args: &[String]) -> Result<()> {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I wouldn't send the entire command line array to this method (`args: &[String]`) because it doesn't seem to be needed. I would only send the value if it exists (`maybe_value: Option<&String>`).

File: fc_util/src/arg_parser.rs [hunk 14]
```diff
@@ -0,0 +1,581 @@
...
+                self.is_valid(arg_info, index, &args)?;
+                let value = if arg_info.takes_value {
+                    ArgParser::next_arg(index + 2, &args)?;
+                    // Safe to unwrap here because we are already checking if there is an argument
+                    // provided at `index + 1` position in `is_valid` call.
+                    args.get(index + 1).unwrap().clone()
+                } else {
+                    ArgParser::next_arg(index + 1, &args)?;
+                    "".to_string()
+                };
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I would do something like:
```
let value = if (args.len() > index + 1 && !args[index + 1].starts_with("--")) {
    Some(args[index + 1])
} else {
    None
};
self.is_valid(arg_info, value)?;
```

File: fc_util/src/arg_parser.rs [hunk 15]
```diff
@@ -0,0 +1,581 @@
...
+        self.optional = optional;
+        self
+    }
+
+    pub fn help(mut self, help: &'a str) -> ArgInfo<'a> {
+        self.help = Some(help);
+        self
+    }
+
+    pub fn update_value(mut self, value: String) -> ArgInfo<'a> {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I would rename this to "set_user_value". Also is it possible to receive a `&'a str` here instead of a `String` ?

File: fc_util/src/arg_parser.rs [hunk 16]
```diff
@@ -0,0 +1,581 @@
...
+            process::exit(0)
+        }
+
+        self.extra_args = if args.contains(&"--".to_string()) {
+            let index = args.iter().position(|r| r.as_str() == "--").unwrap();
+            let (_, extra_args) = args.split_at(index + 1);
+            extra_args.to_vec()
+        } else {
+            Vec::new()
+        };
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I would do this in the `parse()` method. And then call `get_matches()` supplying only the slice with the parameters before "--" .

File: fc_util/src/arg_parser.rs (hunk 15 above)
Comment (serban300) [reply fc_util/src/arg_parser.rs]: I meant holding this struct member (`user_value`) as a &'a str . Just like all the other. I don't know if it makes sense.

File: fc_util/src/arg_parser.rs (hunk 16 above)
Comment (serban300) [reply fc_util/src/arg_parser.rs]: I had the impression that this should be the desired behavior. I thought that we shouldn't parse the parameters supplied after double dash. To me it doesn't make much sense to both parse the parameters after double dash and also add them to an internal struct member "extra_args". Looks like hardcoded application logic to me that fixes a specific jailer corner case.

From what I could gather a double dash is used to signify the end of command options, after which only positional parameters are accepted. We should check what clap does in this situation.

**Comment by lauralt** [comment]
> Overall to me it still feels like the strategy implemented here generates too many corner cases. This makes it hard to follow the code logic and very easy to miss something when doing a change.
> 
> Instead we should rethink the problem and find a strategy that enables us to avoid these special cases and transform them into the generic case. As very well explained here on a simple example: https://www.youtube.com/watch?v=o8NPllzkFhE&t=14m20s if it makes sense.
> 
> Everything revolves around the `for` loop in `get_matches` . We should try to make it as generic as possible.

I've updated the PR with an alternative for that loop. Now it should be easier to follow :D.

File: fc_util/src/arg_parser.rs (hunk 15 above)
Comment (lauralt) [reply fc_util/src/arg_parser.rs]: As per our chat discussion, we've decided to keep `String` here because of the lifetimes conflicts that would appear.

File: fc_util/src/arg_parser.rs [hunk 17]
```diff
@@ -0,0 +1,542 @@
...
+    pub fn parse(self) -> Result<Self> {
+        let mut args: Vec<String> = env::args().collect();
+        // Remove the first element of `args` as it should be the binary name.
+        args.remove(0);
+
+        self.get_matches(args)
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, mut args: Vec<String>) -> Result<Self> {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I think it would be better to receive the args as a reference (`&mut args: Vec<String>`) or slice (`&[String]`)

File: fc_util/src/arg_parser.rs [hunk 18]
```diff
@@ -0,0 +1,542 @@
...
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(self)
+    }
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: There's a lot of logic here. It would be nice to split it in a couple of smaller methods if possible.

File: fc_util/src/arg_parser.rs [hunk 19]
```diff
@@ -0,0 +1,586 @@
...
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    pub fn split(args: &[String]) -> (&[String], &[String]) {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: nit: I would rename this method in order to be more suggestive. `split_args` maybe. Also a comment about what this method does would be very helpful.

File: fc_util/src/arg_parser.rs [hunk 20]
```diff
@@ -0,0 +1,586 @@
...
+    help_header: String,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args_matches.insert(arg_info);
+        self
+    }
+
+    pub fn help(mut self, header: String) -> Self {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: nit: The name of the method should match the name of the field. I would either rename the method to `help_header` or rename the member to `help`. Or maybe rename them both to `info` or `about`. Or define 2 fields like clap: `about` and `version`. Whatever you think is the best fit here.

File: fc_util/src/arg_parser.rs [hunk 21]
```diff
@@ -0,0 +1,586 @@
...
+            process::exit(0)
+        }
+
+        let mut index = 0;
+        while index < args.len() {
+            let arg = &args[index];
+
+            self.is_valid(arg)?;
+
+            let arg_info = self.args_matches.get(&arg[DOUBLE_DASH.len()..]).unwrap();
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: nit: I know it's safe to unwrap, but I would still prefer the `ok_or()` approach.

File: fc_util/src/arg_parser.rs [hunk 22]
```diff
@@ -0,0 +1,586 @@
...
+        self.check_requirements(&args)?;
+
+        Ok(self)
+    }
+
+    // Extracts an argument's value or returns a specific error if the argument is missing.
+    // The unwrap is safe here because we call this function with an `arg_name` parameter
+    // that is for sure contained in `args_matches`.
+    pub fn get_value(&self, arg_name: &'static str) -> Result<String> {
+        if let Some(value) = self.args_matches.get(arg_name).unwrap().value() {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: I'm not sure the unwrap is safe. We should test this with `ok_or() ...` first.

File: jailer/src/env.rs [hunk 23]
```diff
@@ -301,11 +302,12 @@ impl Env {
...
-                .arg(format!("--seccomp-level={}", self.seccomp_level))
-                .arg(format!("--start-time-us={}", self.start_time_us))
-                .arg(format!("--start-time-cpu-us={}", self.start_time_cpu_us))
-                .arg(format!("--api-sock=/{}", socket_file_name))
+                .args(&["--id", &self.id])
+                .args(&["--seccomp-level", &self.seccomp_level.to_string()])
+                .args(&["--start-time-us", &self.start_time_us.to_string()])
+                .args(&["--start-time-cpu-us", &self.start_time_cpu_us.to_string()])
+                .arg("--api-sock")
+                .arg(format!("/{}", socket_file_name))
```
Comment (serban300) [inline jailer/src/env.rs]: nit: are these changes needed ?. Personally I find the previous version more readable.

File: jailer/src/lib.rs [hunk 24]
```diff
@@ -214,122 +215,102 @@ impl fmt::Display for Error {
...
+
+            if fd > 2 {
+                // Safe because close() cannot fail when passed a valid parameter.
+                unsafe { libc::close(fd) };
+            }
+        }
+    }
+}
+
+pub fn set_args_info(arg_parser: ArgParser) -> ArgParser {
```
Comment (serban300) [inline jailer/src/lib.rs]: I wouldn't return an `ArgParser` here. It makes sense to do that as part of builder pattern when the method is defined inside the class. But using it here looks strange to me.

File: fc_util/src/arg_parser.rs [hunk 25]
```diff
@@ -0,0 +1,586 @@
...
+
+    pub fn split(args: &[String]) -> (&[String], &[String]) {
+        if let Some(index) = args.iter().position(|arg| arg == DOUBLE_DASH) {
+            return (&args[0..index], &args[index + 1..args.len()]);
+        }
+
+        (&args[0..args.len()], &[])
+    }
+
+    pub fn parse(mut self) -> Result<Self> {
```
Comment (serban300) [inline fc_util/src/arg_parser.rs]: can this take a reference to self ? `&mut self` . I wouldn't take ownership of the object if not needed. Also I wouldn't return Self here. `Result<()>` should be ok.

File: jailer/src/env.rs (hunk 23 above)
Comment (lauralt) [reply jailer/src/env.rs]: Unfortunately yes, because `clap` accepts also the `--arg=value` option for passing parameters, our implementation does not.

**Comment by dianpopa** [comment]
Can you please test before merging what happens now for [this scenario](https://github.com/firecracker-microvm/firecracker/issues/1082) and update the issue correspondingly?

File: utils/src/arg_parser.rs [hunk 26]
```diff
@@ -0,0 +1,582 @@
...
+            } else {
+                FLAG_PROVIDED
+            };
+            let updated_arg = arg_info.clone().set_user_value(arg_value.to_string());
+
+            let step = if arg_info.takes_value { 2 } else { 1 };
+            index += step;
+
+            self.args_matches.replace(updated_arg);
+        }
```
Comment (dhrgit) [inline utils/src/arg_parser.rs]: A more rusty way of doing the same thing, while avoiding the `clone()` operation, would be:

```rust
        let mut iter = args.iter();
        while let Some(arg) = iter.next() {
            self.is_arg_valid(arg)?;

            let arg_info = self
                .args_matches
                .take(&arg[DOUBLE_DASH.len()..])
                .ok_or_else(|| Error::UnexpectedArgument(arg[DOUBLE_DASH.len()..].to_string()))?;

            let val = if arg_info.takes_value {
                iter.next()
                    .filter(|v| !v.starts_with(DOUBLE_DASH))
                    .ok_or_else(|| Error::MissingValue(arg_info.name.to_string()))?
            } else {
                FLAG_PROVIDED
            };

            self.args_matches.insert(arg_info.set_user_value(val.to_string()));
        }
```

File: utils/src/arg_parser.rs [hunk 27]
```diff
@@ -0,0 +1,582 @@
...
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            optional: false,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> ArgInfo<'a> {
```
Comment (dhrgit) [inline utils/src/arg_parser.rs]: nit: the explicit lifetime is not needed here (nor for the other setters). You can just do:

```rust
pub fn required(mut self, required: bool) -> Self { ... }
```

File: utils/src/arg_parser.rs [hunk 28]
```diff
@@ -0,0 +1,582 @@
...
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The required argument was not provided.
+    MissingArgument(String),
+    /// A value for the argument was not provided.
+    MissingValue(String),
+    /// The provided argument was not expected.
+    UnexpectedArgument(String),
+    /// The argument was provided more than once.
+    DuplicatedArgument(String),
```
Comment (dhrgit) [inline utils/src/arg_parser.rs]: nit: IMO, this should be `DuplicateArgument(String)` (i.e. the arg is a duplicate).

File: utils/src/arg_parser.rs [hunk 29]
```diff
@@ -0,0 +1,582 @@
...
+            return (&args[..index], &args[index + 1..]);
+        }
+
+        (&args, &[])
+    }
+
+    pub fn parse(&mut self) -> Result<()> {
+        let args: Vec<String> = env::args().collect();
+
+        let (args, extra_args) = ArgParser::split_args(&args[1..]);
```
[the remark this answers is not in the document]
Comment (lauralt) [reply utils/src/arg_parser.rs]: Done.

File: utils/src/arg_parser.rs [hunk 30]
```diff
@@ -0,0 +1,582 @@
...
+            }
+            // For the arguments that require a specific argument to be also present in the list
+            // of arguments provided by user, search for that argument.
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (lauralt) [reply utils/src/arg_parser.rs]: I agree it's a bit misleading, but I think it's okay in this case to validate A and to consider B the cause of failure for the parsing and the one that will return an error.

File: utils/src/arg_parser.rs [hunk 31]
```diff
@@ -0,0 +1,582 @@
...
+            .user_value
+            .is_some()
+        {
+            return Err(Error::DuplicatedArgument(arg_name.to_string()));
+        }
+        Ok(())
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(&mut self, args: &[String]) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply utils/src/arg_parser.rs]: I am using it also in a test from `jailer` because `parse()` also collects the arguments and I couldn't use it for testing.

File: utils/src/arg_parser.rs (hunk 26 above)
Comment (lauralt) [reply utils/src/arg_parser.rs]: Looks nice :), changed it.

**Comment by dianpopa** [comment]
@dhrgit your comments are addressed, please re-review.

File: src/utils/src/arg_parser.rs [hunk 32]
```diff
@@ -0,0 +1,567 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::env;
+use std::fmt;
+use std::result;
+
+pub type Result<T> = result::Result<T, Error>;
+
+const ARG_SEPARATOR: &str = "--";
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: The `--` string is used both as a separator (e.g. `arg1 arg2 -- forwarded_arg1`), and as a prefix (e.g. `--arg`). We should make that distinction, and avoid code like `starts_with(ARG_SEPARATOR)` which sounds confusing.

File: src/utils/src/arg_parser.rs [hunk 33]
```diff
@@ -0,0 +1,567 @@
...
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> Self {
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: Please document all public functionality.

File: src/utils/src/arg_parser.rs [hunk 34]
```diff
@@ -0,0 +1,567 @@
...
+            arg_info.user_value = Some(val.to_string());
+        }
+
+        self.validate_matches(&args)?;
+
+        Ok(())
+    }
+
+    // Extracts an argument's value or returns a specific error if the argument is missing.
+    pub fn get_value(&self, arg_name: &'static str) -> Result<String> {
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: I suggest calling this `arg_value()`, and avoid the confusion, since verb member functions are assumed to work on `self`.

File: src/utils/src/arg_parser.rs [hunk 35]
```diff
@@ -0,0 +1,567 @@
...
+            .arg_value()
+            .ok_or_else(|| Error::MissingValue(arg_name.to_string()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn build_parser(arg_parser: ArgParser) -> ArgParser {
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: Why receive the parser as an argument here? It seems that this is always called with `ArgParser::default()`.

File: src/utils/src/arg_parser.rs [hunk 36]
```diff
@@ -0,0 +1,567 @@
...
+
+        // Check if the argument was not provided more than once.
+        if self.args.get(arg_name).unwrap().user_value.is_some() {
+            return Err(Error::DuplicateArgument(arg_name.to_string()));
+        }
+        Ok(())
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(&mut self, args: &[String]) -> Result<()> {
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: A `get_` function implies a returned value, but this function doesn't have one. This function looks like the workhorse of `ArgParser`, while `ArgParser::parse()` deals with minor setup and calls into here.

I don't really understand the use of `matches` either here or with `validate_matches`. Can you please clarify this either by choosing more specific names, or by commenting in a short description of how this all works?

File: src/utils/src/arg_parser.rs [hunk 37]
```diff
@@ -0,0 +1,567 @@
...
+    // Collects the command line arguments and the values provided for them.
+    pub fn parse(&mut self) -> Result<Help> {
+        let args: Vec<String> = env::args().collect();
+
+        // Skipping the first element of `args` as it is the name of the binary, not an actual argument.
+        let (args, extra_args) = ArgParser::split_args(&args[1..]);
+        self.extra_args = extra_args.to_vec();
+
+        // If user provided the `--help` argument, we want to stop the parsing.
+        if args.contains(&HELP_ARG.to_string()) {
```
Comment (dhrgit) [inline src/utils/src/arg_parser.rs]: The user should decide on how to treat any args, including `--help`, and the parser should continue to do its job, not stop here.

Consider, for instance, a subcommand scenario like `command  subcommand --help`. In this case, the user might choose to show specific help information regarding `subcommand`, so they might need the whole command line parsed (i.e. to identify the subcommand).

The parser should just parse the information and validate it against the provided schema. It's not its job to interpret the meaning of the information it parses.

File: src/utils/src/arg_parser.rs [hunk 38]
```diff
@@ -0,0 +1,748 @@
...
+    }
+}
+
+/// Represents the value of an argument, which will be a `String` if
+/// the argument takes a value, or `bool` if it's a flag.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Value {
+    Bool(bool),
+    String(String),
+}
```
[the remark this answers is not in the document]
Comment (serban300) [reply src/utils/src/arg_parser.rs]: On a second thought a better suggestion would be to use something like this:

```
pub struct ArgInfo<'a> {
    name: &'a str,
    required: bool,
    requires: Option<&'a str>,
    help: Option<&'a str>,

    value_info: ValueInfo,
}

pub enum ValueInfo {
    Flag(FlagValueInfo),
    Input(InputValueInfo),
}

pub struct FlagValueInfo {
    default_value: bool,
    user_value: bool,
}

pub struct InputValueInfo {
    default_value: Option<String>,
    user_value: Option<String>,
}
```

This would eliminate the need for the `takes_value`, and would ensure that `ArgInfo` doesn't contain invalid data. For example not you can have an `ArgInfo` instance with the following data:

```
ArgInfo<'a> {
    ...
    takes_value: false,
    default_value: "value",
    user_value: "x",
}
```

or

```
ArgInfo<'a> {
    ...
    takes_value: true,
    default_value: true,
    user_value: "x",
}
```

But it might be a lot of work to implement and it might not be worth it. So feel free to discard the suggestion.

[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/utils/src/arg_parser.rs]: I find it more clear as it is right now. We duplicate the information if both boolean and values that take values have separate structures with the same fields.

I would propose to iterate over this in a next PR because there are still a few tiny things we need to fix anyway.

File: src/firecracker/src/main.rs [hunk 39]
```diff
@@ -94,105 +78,108 @@ fn main() {
...
-    let bind_path = cmd_arguments
-        .value_of("api_sock")
+                .default_value(DEFAULT_API_SOCK_PATH)
+                .help("Path to unix domain socket used by the API"),
+        ).arg(
+        ArgInfo::new("id")
+            .takes_value(true)
+            .default_value(DEFAULT_INSTANCE_ID)
+            .help("MicroVM unique identifier"),
+    ).arg(
```
Comment (andreeaflorescu) [inline src/firecracker/src/main.rs]: The formatting here looks wrong and `cargo fmt` doesn't seem to catch this error. Can you manually format it please?

File: src/utils/src/arg_parser.rs (hunk 38 above)
[the remark this answers is not in the document]
Comment (serban300) [reply src/utils/src/arg_parser.rs]: Ok. Sounds good to me.
